source/exe/win32/service_base.cc

```diff
@@ -0,0 +1,180 @@
+  service_static = &service;
+
+  // The `SERVICE_TABLE_ENTRY` struct requires a volatile `LPSTR`
+  char nullstr[1] = "";
+  SERVICE_TABLE_ENTRYA service_table[] = {// Even though the service name is ignored for own process
+                                          // services, it must be a valid string and cannot be 0.
+                                          {nullstr, (LPSERVICE_MAIN_FUNCTIONA)ServiceMain},
+                                          // Designates the end of table.
+                                          {0, 0}};
+
+  if (!::StartServiceCtrlDispatcherA(service_table)) {
+    auto last_error = ::GetLastError();
+    // There are two cases:
+    // 1. The user is trying to run Envoy as a console app. In that the last error is
+    // `ERROR_FAILED_SERVICE_CONTROLLER_CONNECT`. In that case we return false and let Envoy start
+    // as normal.
+    // 2. Other programmatic errors.
+    if (last_error == ERROR_FAILED_SERVICE_CONTROLLER_CONNECT) {
+      return false;
+    } else {
+      PANIC(
+          fmt::format("Could not dispatch Envoy to start as a service with error {}", last_error));
+    }
+  }
+  return true;
+}
+
+DWORD ServiceBase::Start(std::vector<std::string> args) {
+  // Run the server listener loop outside try/catch blocks, so that unexpected exceptions
+  // show up as a core-dumps for easier diagnostics.
+  absl::InitializeSymbolizer(args[0].c_str());
+  std::shared_ptr<Envoy::MainCommon> main_common;
+
+  // Initialize the server's main context under a try/catch loop and simply return `EXIT_FAILURE`
+  // as needed. Whatever code in the initialization path that fails is expected to log an error
+  // message so the user can diagnose.
+  try {
+    main_common = std::make_shared<Envoy::MainCommon>(args);
+    Envoy::Server::Instance* server = main_common->server();
+    if (!server->options().signalHandlingEnabled()) {
+      // This means that the Envoy has not registered `ENVOY_SIGTERM`.
+      // we need to manually enable it as we are going to use it to
+      // handle close requests from `SCM`.
+      sigterm_ = server->dispatcher().listenForSignal(ENVOY_SIGTERM, [server]() {
+        ENVOY_LOG_MISC(warn, "caught ENVOY_SIGTERM");
+        server->shutdown();
+      });
+    }
+  } catch (const Envoy::NoServingException& e) {
+    return S_OK;
+  } catch (const Envoy::MalformedArgvException& e) {
+    return E_INVALIDARG;
+  } catch (const Envoy::EnvoyException& e) {
+    ENVOY_LOG_MISC(warn, "Envoy failed to start with {}", e.what());
+    return E_FAIL;
+  }
+
+  return main_common->run() ? S_OK : E_FAIL;
+}
+
+void ServiceBase::Stop(DWORD control) {
+  auto handler = Event::eventBridgeHandlersSingleton::get()[ENVOY_SIGTERM];
+  if (!handler) {
+    PANIC("No handler is registered to stop gracefully, aborting the program.");
+  }
+
+  char data[] = {'a'};
+  Buffer::RawSlice buffer{data, 1};
+  auto result = handler->writev(&buffer, 1);
+  RELEASE_ASSERT(result.rc_ == 1,
```

Comment:
@davinci26 how come we are using `RELEASE_ASSERT` here for error handling? This doesn't match error handling policy in https://github.com/envoyproxy/envoy/blob/main/STYLE.md#error-handling I believe. CC @asraa

Reply:
Or maybe it's fine because you are on the way out anyway? I'd be interested in your take @davinci26 on all the RELEASE_ASSERTs in this module relative to the policy. I actually think it might be fine, but it's hard to tell without understanding the Windows API expectations.

Reply:
> Or maybe it's fine because you are on the way out anyway?

Exactly, we have already received a SIGTERM from the service control manager which means that for some reason we need to shutdown anyway. We try to do it in a graceful way but if the write fails we might as well crash.

The rest of the release asserts are guarding against a developer shooting themselves at the foot. i.e. someone manually calling delete the pointers on the global scope. I added the release asserts for them based on `or an obvious crash on a subsequent line via null pointer dereference.`. No strong opinion though. 

Do you want me to document these in code with comments so it's clear for anyone reading the code?